# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---
#### Δημήτρης Δήμος - 031 17 165
#### Χρήστος Δημόπουλος - 031 17 037
###### 7ο Εξάμηνο - Δεκέμβριος 2020  
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.

Στα δεδομένα της άσκησης περιλαμβάνονται ένα αρχείο με το όνομα movies_metadata.csv το όποιο περιέχει τα χαρακτηριστικά κάθε ταινίας όπως θέμα, σκηνοθέτης ηθοποιοί, λέξεις κλειδιά κ.α. από το imdb καθώς και τα αρχεία ratings.csv τα όποια περιέχουν πραγματικές αξιολογήσεις χρηστών, χωρισμένες σε train και σε test.

# Εκφώνηση
Στην παρούσα εργασία σας ζητείται να μελετήσετε και να υλοποιήσετε τα παρακάτω:<br>

### Μέρος 1
Στο μέρος 1 και 2 θα ασχοληθείτε και θα εργαστείτε μόνο με το αρχείο movies_metadata.csv.
#### Ερώτημα 1α

Αρχικά θα πρέπει, αφού μελετήσετε τη δομή και τα χαρακτηριστικά του αρχείου αυτού, να κατασκευάσετε μια βάση γνώσης για την Prolog η όποια ουσιαστικά θα αποτελεί τον κόσμο με τον όποιο θα εργαστείτε στην συνέχεια. Τα κατηγορήματα που θα δημιουργηθούν θα σας βοηθήσουν και στην κατασκευή του recommender και θα είναι της μορφής:

```
director(Movie, Director).
genre(Movie, Genre).
```
#### Ερώτημα 1β

Αφού δημιουργήσετε τον κόσμο του προβλήματος, στη συνέχεια καλείστε να δημιουργήσετε, σε Prolog, απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες

Αξίζει να σημειωθεί ότι στα παραπάνω μπορείτε να προσθέσετε περισσότερα ερωτήματα ή να αλλάξετε την κλιμακωσιμότητα των queries (πέρα από το ίδιο, αρκετά ίδιο, σχετικά ίδιο) που θα κατασκευάσετε(π.χ. μια κλίμακα από 1 στα 5 όπου αυτό είναι δυνατόν), μιας και αυτά στην συνέχεια θα χρησιμοποιηθούν για την λειτουργία του recommender. Έτσι μπορείτε να προσθέσετε queries με τα όποια μπορεί να παράγονται καλύτερες συστάσεις  (το αρχείο movie_metadata.csv περιέχει πολλές πληροφορίες ακόμα για κάθε ταινία όπως έτος κυκλοφορίας, βαθμολογία στο imdb, facebοok_likes κ.α.).

Περισσότερες πληροφορίες για το dataset μπορείτε να διαβάσετε σε αυτό το [Link](https://www.kaggle.com/georgefila/movies-metadata):
### Μέρος 2: Recommendation System
Στο σημείο αυτό καλείστε με βάση αυτά που κάνατε στο μέρος 1 να κατασκευάσετε queries τα όποια θα σας επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες. Τα ερωτήματα αυτά θα είναι κλιμακούμενα, δηλαδή θα υπάρχουν ερωτήματα που επιστρέφουν αρκετά κοινές ταινίες αλλά και που επιστρέφουν λιγότερο και λιγότερο κοινές (σε μια κλίμακα π.χ. από 1 σε 5). Για παράδειγμα:

```
find_simmilar_movies_5("Pirates Of The Caribbean", M).
M = "Pirates Of The Caribbean: On Stranger Tides"
M = "The Chronicles Of Narnia"
M = "Prince Of Persia: The Sands Of Time"
...
```
Για παράδειγμα, το παραπάνω ερώτημα θα επιστρέφει αρκετά κοινές σε περιεχόμενο ταινίες με την ταινια "Pirates Of The Caribbean". Θα υπάρχουν και αντίστοιχα ερωτήματα που θα βρίσκουν λιγότερο όμοιες ταινίες. Ο δείκτης ομοιότητας των ταινιών είναι αυθαίρετος και μπορείτε να τον ορίσετε εσείς όπως θέλετε, αρκεί να υπάρχει κάποια λογική σύνδεση με τα δεδομένα που περιέχονται στο αρχείο movies_metadata.csv. 

Συνεπώς η συνάρτηση που θα κάνει τις προτάσεις (recommendation) με είσοδο μια ταινία πρέπει να επιστρέφει-εκτυπώνει μια λίστα με τις προτεινόμενες ταινίες κατά φθίνουσα σειρά ομοιότητας. 


### Μέρος 3: Recommendation System Με βάση τις προτιμήσεις - Αξιολογήσεις του χρήστη


Σε αυτό το σημείο θα εργαστείτε με τα αρχεία ratings τα όποια περιέχουν αξιολογήσεις (από 1 μέχρι 5) για τις παραπάνω ταινίες. Το προηγούμενο σύστημα συστάσεων προτείνει στον χρήστη ταινίες αποκλειστικά με βάση την ομοιότητά τους. Σε αυτό το σημείο θα γίνει μια αναβάθμιση του συστήματος έτσι ώστε να παράγονται καλύτερες  συστάσεις οι όποιες θα λαμβάνουν υπόψιν και τις προτιμήσεις του χρήστη, οι όποιες θα εξάγονται από τις αξιολογήσεις που έχει κάνει μέχρι στιγμής. 

Η εκπαίδευση του recommender θα γίνεται ως εξής:

Για κάθε ταινία θα υπάρχει ένα score το όποιο αρχικά θε είναι ίσο με 0 και θα διαμορφώνεται από τις αξιολογήσεις κάθε user. Έτσι για έναν χρήστη με βάση τις αξιολογήσεις που υπάρχουν στο αρχείο train_ratings θα πρέπει:

1.	Για κάθε ταινία που έχει βαθμολογήσει να βρίσκονται οι κοινές ταινίες ανά κλίμακα και στο μέχρι τώρα σκορ κάθε παρόμοιας ταινίας θα προστίθεται ένα βάρος το όποιο θα μπορούσε να είναι το ποσοστό ομοιότητας της ταινίας (δηλαδή ένα βάρος για κάθε κλίμακα) επί τον βαθμό που έχει βάλει ο χρήστης για την αρχική ταινία. Στη συνέχεια, ανάλογα με το σκορ που έχει σχηματιστεί για κάθε ταινία, θα επιλέγεται αν αυτή θα μπορούσε να είναι προτεινόμενη για τον χρήστη.

Η λογική πίσω από την παραπάνω διαδικασία είναι ότι παρόμοιες ταινίες θα έχουν ανάλογο βαθμό. Για παράδειγμα αν ένας χρήστης έχει αξιολογήσει αρκετές  ταινίες οι όποιες είναι sci-fiction με 5/5 τότε μια ταίνια sci-fiction την όποια δεν έχει δεί λογικά θα του αρέσει και θα έπρεπε να την προτείνουμε.

Μετά την εκπαίδευση του συστήματος σας καλείστε να δοκιμάσετε τον recommender που κατασκευάσατε στην πράξη. Για τον σκοπό αυτό θα φορτώσετε το αρχείο train_ratings.csv όπου περιέχονται οι αξιολογήσεις του ίδιου χρήστη για άλλες ταινίες. Το σύστημά σας πρέπει να προβλέπει αν μια ταινία θα πρέπει να προταθεί στον χρήστη. Μια ταίνια θα έπρεπε να έχει προταθεί στον  χρήστη, αν έχει βαθμό μεγαλύτερο του 3. Συνεπώς για την επίβλεψη του συστήματός σας θα πρέπει για κάθε μια από τις ταινίες του αρχείου test_ratings.csv να επιστρέφεται 1 ή 0 αν η ταινία θα έπαιρνε βαθμό μεγαλύτερο του 3 ή όχι, δηλαδή αν θα έπρεπε να την είχαμε προτείνει ή όχι.

Στην συνέχεια, σε συνδιασμό με τις πραγματικές απαντήσεις του χρήστη θα αξιολογήσετε το σύστημά σας χρησιμοποιόντας τις μετρικές: precision, recall, f1 οι οποίες είναι οι πλέον γνωστές μετρικές και ευρέως χρησιμοποιούμενες τεχνικές για την επίβλεψη-μέτρηση απόδοσης ανάλογων συστημάτων.

1. Precision: Δείχνει πόσο ακριβές είναι το σύστημα. Υπολογίζει πόσα από τα στιγμιότυπα τα όποια προβλέψαμε ότι ανήκουν σε μια κλάση όντως ανήκουν σε αυτή. Η μετρική αυτή μας δίνει μια εικόνα σχετικά με τον αριθμό των ταινίων που προβλέψαμε ως προτεινόμενες ενώ δεν θα έπρεπε.

2. Recall: Υπολογίζει πόσα από τα στιγμιότυπα που ανήκουν σε μια κλάση (π.χ. προτεινόμενες ταινίες) προβλέφθηκαν σωστά.

3. F1: Είναι ένας μέσος μεταξύ των παραπάνω δυο μετρικών, έτσι ώστε να διατηρείται μια ισορροπία μεταξύ τους. Υπολογίζεται από την παρακάτω σχέση:

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

Οι παραπάνω συναρτήσεις παρέχονται από την βιβλιοθήκη scikit-learn.

Τέλος για την καλύτερη επίβλεψη του συστήματος σας μπορείτε να εκπαιδεύσετε  τον recommender σας με ένα υποσύνολο ταινίων (π.χ. 10, 20, 50, ...) για να μελετήσετε κατά πόσο σας βοηθούν οι επιπλέον αξιολογήσεις κάθε φορά (δηλαδή κατά πόσο βελτιώνονται οι παραπάνω μετρικές στο test set). Έτσι π.χ. μπορείτε να εντοπίσετε περιπτώσεις όπως για παράδειγμα ότι με έναν recommender μπορεί να μην επιτυγχάνετε πολύ υψηλό σκορ όσο με άλλους, αλλά το βέλτιστο σκόρ σας επιτυγχάνεται πολύ γρηγόρα π.χ. με μόνο 10 ταινίες αντί 100.



# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


In [1]:
import pandas as pd   # Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
from pyswip import Prolog
import os

In [2]:
path = os.getcwd()   # το αρχείο 'movie_metadata.csv' βρίσκεται στο ίδιο directory με το παρόν script

# Διάβασμα του αρχείου 'movie_metadata.csv' 
data = pd.read_csv(path + "/movies_metadata.csv") 

# Στο csv υπαρχούν κελία με nan τιμές. Στις θέσεις αυτές βάζουμε 'UNK' με τη συνάρτηση:
data.fillna("UNK", inplace=True)

# Preview the first 5 lines of the loaded data 
data.head()

Unnamed: 0     budget                                    genres  \
0           0  237000000  Action|Adventure|Fantasy|Science Fiction   
1           1  300000000                  Adventure|Fantasy|Action   
2           2  245000000                    Action|Adventure|Crime   
3           3  250000000               Action|Crime|Drama|Thriller   
4           4  260000000          Action|Adventure|Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                       plot_keywords  language  \
0  culture clash|future|space war|space colony|so...   English   
1  ocean|drug abuse|exotic island|east india trad...   English   
2  spy|based on novel|secret agent|sequel|mi6|bri...  Français   
3  dc comics|crime fighter|terrorist|secret ident...   English   
4  based on novel|mars|medallion|space travel|pri...   English   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ...  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   
3  Following the death of District Attorney Harve...  112.312950  ...   
4  John Carter is a war-weary, former military ca...   43.926995  ...   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                movie_title vote_average  num_voted_users  \
0                                    Avatar          7.2            11800   
1  Pirates of the Caribbean: At World's End          6.9             4500   
2                                   Spectre          6.3             4466   
3                     The Dark Knight Rises          7.6             9106   
4                               John Carter          6.1             2124   

  title_year                   country      director_name     actor_1_name  \
0       2009  United States of America      James Cameron      Zoe Saldana   
1       2007  United States of America     Gore Verbinski    Orlando Bloom   
2       2015            United Kingdom         Sam Mendes  Christoph Waltz   
3       2012  United States of America  Christopher Nolan    Michael Caine   
4       2012  United States of America     Andrew Stanton     Lynn Collins   

       actor_2_name       actor_3_name  
0  Sigourney Weaver       Stephen Lang  
1   Keira Knightley  Stellan Skarsgård  
2       Léa Seydoux      Ralph Fiennes  
3       Gary Oldman      Anne Hathaway  
4   Samantha Morton       Willem Dafoe  

[5 rows x 27 columns]

In [3]:
def clean_text(text):
    text = text.replace(u'\xa0', u'')
    text = text.replace(u"'", u'')
    return text

---
### Ερώτημα 1α - Κατασκευή Κόσμου

Στην αρχή της άσκησης εισάγουμε στη βάση γνώσης τα απαραίτητα κατηγορήματα. Επιλέγουμε να εισάγουμε μόνο στοιχεία που θα μπορούσαν να μας χρησιμεύσουν εκ πρώτης όψεος στην κατασκευή ενός recommender. Εισάγουμε, λοιπόν, τα εξής:

1. movie(Movie)
2. genre(Movie, Genre)
3. director(Movie, Director)
4. plot_keyword(Movie, Key)
5. actor_1_name(Movie, Actor)
6. actor_2_name(Movie, Actor)
7. actor_3_name(Movie, Actor)
8. actor(Movie, Actor) 
9. language(Movie, Language)

Το κατηγόρημα 8 προστέθηκε στο τέλος, καθώς έγινε σε μετέπειτα στάδιο αντιληπτό πως δεν είναι απαραίτητη η γνώση του κατά πόσο πρωταγωνιστεί ο ηθοποιός στην αντίστοιχη ταινία.

---

In [4]:
# Ορίζουμε τον κόσμο μας
prolog = Prolog()

# Για κάθε row του πίνακα φτιάχνουμε κατηγορήματα και τα αποθηκεύουμε σε μια λίστα literals
literals = []
movie_score = {}
for row in data.itertuples(index=True, name='Pandas'):
    movie_title = clean_text(getattr(row, 'movie_title'))
    director = clean_text(getattr(row, 'director_name'))
    actor_1_name = clean_text(getattr(row, 'actor_1_name'))
    actor_2_name = clean_text(getattr(row, 'actor_2_name'))
    actor_3_name = clean_text(getattr(row, 'actor_3_name'))
    language = clean_text(getattr(row, 'language'))
    
    literals.append("movie('"+ movie_title +"')")
    
    # genre
    for genre in getattr(row, 'genres').split("|"):
        literals.append("genre('"+ movie_title +"','"+ genre +"')")
        
    # director
    literals.append("director('"+ movie_title +"','"+ director +"')")
    
    # plot
    for plot_keyword in getattr(row, 'plot_keywords').split("|"):
        plot_keyword = clean_text(plot_keyword)
        literals.append("plot_keyword('"+ movie_title +"','"+ plot_keyword +"')")
        
    # actors
    literals.append("actor_1_name('"+ movie_title +"','"+ actor_1_name +"')")
    literals.append("actor_2_name('"+ movie_title +"','"+ actor_2_name +"')")
    literals.append("actor_3_name('"+ movie_title +"','"+ actor_3_name +"')")
    literals.append("actor('"+ movie_title +"','"+ actor_1_name +"')")
    literals.append("actor('"+ movie_title +"','"+ actor_2_name +"')")
    literals.append("actor('"+ movie_title +"','"+ actor_3_name +"')")
    
    # language
    literals.append("language('"+ movie_title +"','"+ language +"')")
    

# ταξινομούμε τα κατηγορήματα πριν τα εισάγουμε στη βάση γνώσης 
literals.sort()
for literal in literals:
    prolog.assertz(literal)

---
### Ερώτημα 1β - Κατασκευή Κανόνων

Κατασκευάζουμε κανόνες, ο κάθε ένας εκ των οποίων περιγράφεται με σχόλιο. Στην αρχή κάποιοι κανόνες είχαν κατασκευαστεί, ώστε να είναι πιο αυστηροί (πχ. αντί για "τουλάχιστον 2 κοινά genre" υπήρχε το "ακριβώς 2 κοινά genre") αλλά, εν τέλει, θεωρήθηκαν πιο εύχρηστοι οι πιο ελαστικοί κανόνες. Ο λόγος θα φανεί στη συνέχεια.

---

In [5]:
predicates = ["","","","","","","","","","","","","",""]

# 0. Βοηθητικό Κατηγόρημα για να μην προκύπτουν απαντήσεις επειδή ορισμένες ταινίες έχουν κοινό το UNK
predicates[0] = 'not_UNK(X) :- X \= \'UNK\''

# 1. Κοινό θέμα - τουλάχιστον 5 κοινά genre
predicates[1] = 'common_5_genre(X,Y) :- genre(X,G1), genre(Y,G1), genre(X,G2), genre(Y,G2), genre(X,G3), genre(Y,G3), genre(X,G4), genre(Y,G4), genre(X,G5), genre(Y,G5), X \= Y, G1 \= G2, G1 \= G3, G1 \= G4, G1 \= G5, G2 \= G3, G2 \= G4, G2 \= G5, G3 \= G4, G3 \= G5, G4 \= G5, not_UNK(G1), not_UNK(G2), not_UNK(G3), not_UNK(G4), not_UNK(G5)'

# 2. Αρκετά κοινό θέμα - τουλάχιστον 3 κοινά genre
predicates[2] = 'common_3_genre(X,Y) :- genre(X,G1), genre(Y,G1), genre(X,G2), genre(Y,G2), genre(X,G3), genre(Y,G3), X \= Y, G1 \= G2, G1 \= G3, G2 \= G3, not_UNK(G1), not_UNK(G2), not_UNK(G3)'
         
# 3. Tουλάχιστον 2 κοινά genre   
predicates[3] = 'common_2_genre(X,Y) :- genre(X,G1), genre(Y,G1), genre(X,G2), genre(Y,G2), X \= Y, G1 \= G2, not_UNK(G1), not_UNK(G2)'    
    
# 4. Σχετικά κοινό θέμα - ακτιβώς 1 κοινό genre
predicates[4] = 'common_1_genre(X,Y) :- genre(X,G1), genre(Y,G1), X \= Y, not_UNK(G1), (genre(X,G2), genre(Y,G2) -> G2 = G1)'

# 5. Κοινός σκηνοθέτης
predicates[5] = 'common_director(X,Y) :- director(X,Dir), director(Y,Dir), X \= Y, not_UNK(Dir)'

# 6. Ακριβώς ίδια πλοκή - τουλάχιστον 4 κοινά plot_keywords
predicates[6] = 'common_4_plot_keyword(X,Y) :- plot_keyword(X,Key1), plot_keyword(X,Key2), plot_keyword(X,Key3), plot_keyword(X,Key4), plot_keyword(Y,Key1), plot_keyword(Y,Key2), plot_keyword(Y,Key3), plot_keyword(Y,Key4), X \= Y, Key1 \= Key2, Key1 \= Key3, Key1 \= Key4, Key2 \= Key3, Key2 \= Key4, Key3 \= Key4, not_UNK(Key1), not_UNK(Key2), not_UNK(Key3), not_UNK(Key4)'

# 7. Σχετικά ίδια πλοκή - τουλάχιστον 2 κοινά plot_keywords
predicates[7] = 'common_2_plot_keyword(X,Y) :- plot_keyword(X,Key1), plot_keyword(X,Key2), plot_keyword(Y,Key1), plot_keyword(Y,Key2), X \= Y, Key1 \= Key2, not_UNK(Key1), not_UNK(Key2)'

# 8. Τουλάχιστον 1 κοινή plot_keyword
predicates[8] = 'common_1_plot_keyword(X,Y) :- plot_keyword(X,Key), plot_keyword(Y,Key), X \= Y'

# 9. Ίδιους τους βασικούς ηθοποιούς - τουλάχιστον 3 κοινούς actors (και τους 3 δηλαδή)
predicates[9] = 'common_3_actor(X,Y) :- actor_1_name(X,Actor1), actor_2_name(X,Actor2), actor_3_name(X,Actor3), actor_1_name(Y,Actor1), actor_2_name(Y,Actor2), actor_3_name(Y,Actor3), X \= Y, not_UNK(Actor1), not_UNK(Actor2), not_UNK(Actor3)'

# 10. Αρκετά ίδιους βασικούς ηθοποιούς - τουλάχιστον 2 κοινούς actors
predicates[10] = 'common_2_actor(X,Y) :- actor(X,Actor1), actor(X,Actor2), actor(Y,Actor1), actor(Y,Actor2), X \= Y, Actor1 \= Actor2, not_UNK(Actor1), not_UNK(Actor2)'

# 11. Σχετικά ίδιους ηθοποιούς - ακριβώς 1 κοινό actor
predicates[11] = 'common_1_actor(X,Y) :- actor(X,Actor1), actor(Y,Actor1), X \= Y, not_UNK(Actor1), ((actor(X,Other),actor(Y,Other))->Other=Actor1)'

# 12. Ίδια γλώσσα
predicates[12] = 'common_language(X,Y) :- language(X,Z), language(Y,Z), X \= Y, not_UNK(Z)'

# 13. Είναι έγχρωμες ή ασπρόμαυρες
predicates[13] = 'black_n_white(X) :- plot_keyword(X,\'black and white\')'

for predicate in predicates:
    if predicate != "":
        prolog.assertz(predicate)

In [6]:
# Σε αυτό το σημείο αποθηκεύουμε τα δεδομένα της βάσης και σε prolog αρχείο
# για να μπορεί να γίνει ευκολότερο debugging σε prolog interpreter

file = open("db.pl", "w")

for literal in literals:
    file.write(literal)
    file.write(".\n")
    
for predicate in predicates:
    if predicate != "":
        file.write(predicate)
        file.write(".\n")
file.close()

# **2ο Μέρος: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**

Στο σημείο αυτό με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1 θα κατασκευαστούν κατηγορήματα για την ομοιότητα ταινιών. Παρακάτω δίνεται ένα μικρό παράδειγμα ενός κανόνα και πώς αυτός θα μπορούσε να γραφτεί μέσω του Pyswip. Επίσης όπως αναφέρεται και σε σχόλιο παραπάνω θα μπορεί να γραφτεί και μια βάση δεδομένων με τους κανόνες και να γίνει απευθείας consult.

Στο παρακάτω παράδειγμα το 5 και το 4 εκφράζουν την ομοιότητα των ταινιών π.χ. οι ταινίες που παράγονται μέσω του find_similar_5 είναι πιο όμοιες από αυτές που παράγονται μέσω του find_similar_4. 


---
Δημιουργούμε 5 queries. Το καθένα επιστρέφει ταινίες σχετικές με την ταινία εισόδου στον αντίστοιχο βαθμό.

(τα queries αρχικά διαμορφώθηκαν με τρόπο που παρήγαγε καλύτερες τιμές στις μετρικές του τελευταίου ερωτήματος. Παρόλαυτα, ο kernel πέθαινε λόγω φόρτου, οπότε και κρατήσαμε τα παρακάτω queries που ο kernel μπορεί αν αντέξει και είναι αρκετά ικανοποιητικά)

---

In [7]:
# Similarity queries

levels = ["","","","","",""]

# Similarity Level: 1
levels[1] = 'find_sim_1(X,Y) :- common_1_genre(X,Y), X \= Y'

# Similarity Level: 2
levels[2] = 'find_sim_2(X,Y) :- common_language(X,Y), common_1_genre(X,Y), common_1_plot_keyword(X,Y), X \= Y'

# Similarity Level: 3
levels[3] = 'find_sim_3(X,Y) :- common_language(X,Y), common_1_genre(X,Y), common_1_actor(X,Y), common_1_plot_keyword(X,Y), X \= Y'

# Similarity Level: 4
levels[4] = 'find_sim_4(X,Y) :- common_language(X,Y), common_2_genre(X,Y), common_1_actor(X,Y), common_1_plot_keyword(X,Y), X \= Y'

# Similarity Level: 5
levels[5] = 'find_sim_5(X,Y) :- common_language(X,Y), common_2_genre(X,Y), common_2_actor(X,Y), common_2_plot_keyword(X,Y), common_director(X,Y), X \= Y'

In [8]:
for level in levels:
    if level != "":
        prolog.assertz(level)

In [9]:
# Ανανεώνουμε το αρχείο prolog για ευκολότερο debugging

file = open("db.pl", "w")

for literal in literals:
    file.write(literal)
    file.write(".\n")
    
for predicate in predicates:
    if predicate != "":
        file.write(predicate)
        file.write(".\n")
        
for level in levels:
    if level != "":
        file.write(level)
        file.write(".\n")        
file.close()

In [10]:
# Ο simple_recommender λειτουργεί ως εξής:
# 1. Δέχεται μια ταινία ως είσοδο
# 2. Για κάθε επίπεδο ομοιότητας (από το 5 προς μικρότερα) τρέχει το αντίστοιχο query
# 3. Αποθηκεύει τις απαντήσεις-ταινίες σε ένα set
# 4. Φροντίζει να μην τοποθετεί απαντήσεις-ταινίες στο σύνολο,
#    αν αυτές έχουν ήδη βρεθεί ως απάντηση σε προηγούμενο query
# 5. Επιστρέφει έναν πίνακα που περιέχει τα sets


def simple_recommender(movie):
    s1 = set()
    s2 = set()
    s3 = set()
    s4 = set()
    s5 = set()
    
    q1 = prolog.query("find_sim_1('" + movie +"',M)")
    q2 = prolog.query("find_sim_2('" + movie +"',M)")
    q3 = prolog.query("find_sim_3('" + movie +"',M)")
    q4 = prolog.query("find_sim_4('" + movie +"',M)")
    q5 = prolog.query("find_sim_5('" + movie +"',M)")
    
    for soln in q5:
        m = soln['M'] 
        if m not in s5:
            s5.add(soln['M'])
    q5.close()
    
    for soln in q4:
        m = soln['M'] 
        if m not in s4 and m not in s5:
            s4.add(soln['M'])
    q4.close()
    
    for soln in q3:
        m = soln['M'] 
        if m not in s3 and m not in s4 and m not in s5:
            s3.add(soln['M'])
    q3.close()
    
    for soln in q2:
        m = soln['M'] 
        if m not in s2 and m not in s3 and m not in s4 and m not in s5:
            s2.add(soln['M'])
    q2.close()
    
    for soln in q1:
        m = soln['M'] 
        if m not in s1 and m not in s2 and m not in s3 and m not in s4 and m not in s5:
            s1.add(soln['M'])
    q1.close()
    
    answers = ["","","","",""]
    answers[0] = s1
    answers[1] = s2
    answers[2] = s3
    answers[3] = s4
    answers[4] = s5
    return answers

In [11]:
# Παράδειγμα λειτουργίας του simple_recommender για την ταινία "The Hunger Games: Mockingjay - Part 1"

print("Similarity Level: 5")
print(list(simple_recommender('The Hunger Games: Mockingjay - Part 1')[4])[:5])
print()
print("Similarity Level: 4")
print(list(simple_recommender('The Hunger Games: Mockingjay - Part 1')[3])[:5])
print()
print("Similarity Level: 3")
print(list(simple_recommender('The Hunger Games: Mockingjay - Part 1')[2])[:5])
print()
print("Similarity Level: 2")
print(list(simple_recommender('The Hunger Games: Mockingjay - Part 1')[1])[:5])
print()
print("Similarity Level: 1")
print(list(simple_recommender('The Hunger Games: Mockingjay - Part 1')[0])[:5])

Similarity Level: 5
['The Hunger Games: Catching Fire', 'The Hunger Games: Mockingjay - Part 2']

Similarity Level: 4
['The Hunger Games', 'Cirque du Freak: The Vampires Assistant', 'A Scanner Darkly']

Similarity Level: 3
[]

Similarity Level: 2
['Dune', '20,000 Leagues Under the Sea', 'Rise of the Planet of the Apes', 'Sin City: A Dame to Kill For', 'Superman III']

Similarity Level: 1
['Chill Factor', 'VeggieTales: The Pirates Who Dont Do Anything', 'Virgin Territory', 'Nancy Drew', 'Switchback']


# **3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**


Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.




In [12]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score

In [13]:
path = os.getcwd()   # το αρχείο 'train_ratings.csv' βρίσκεται στο ίδιο directory με το παρόν script

data = pd.read_csv(path + "/train_ratings.csv") 
data.fillna("UNK", inplace=True)
data.head(10)

Unnamed: 0  userId  movieId            movie_title  rating   timestamp
0       78477     547      235            Stand by Me     4.0   974780707
1       78747     547     1251  Letters from Iwo Jima     5.0   974809897
2       78808     547     1391      Y Tu Mamá También     0.5  1076809696
3       78584     547      824          Moulin Rouge!     2.0  1076968814
4       80322     547    51540        Horrible Bosses     4.0  1192372083
5       78554     547      581     Dances with Wolves     5.0   974779641
6       78933     547     2043    Along Came a Spider     3.5  1199574304
7       78573     547      745        The Sixth Sense     2.0   974778219
8       78866     547     1729  The Forbidden Kingdom     5.0   986695295
9       78536     547      509           Notting Hill     1.0   986694295

In [14]:
rating_weights = {0: -1, 1: -0.5, 2:0, 3:0, 4:0.5, 5:1}
score_weights = [1, 2, 3, 4, 5]

# Η συνάρτηση train_recommender εκπαιδεύει τον recommender βάσει των ratings χρήστη
def train_recommender(ratings, rating_weights, score_weights, start=0, end = -1, movie_score = {}):
    
    if end == -1:
        end = len(ratings)

    ratings = ratings.iloc[range(start, end)]
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies = simple_recommender(movie)
        
        for i in range(5):
            for similar_movie in list(similar_movies[i]):
                if similar_movie not in movie_score:
                    movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[i]
                else:
                    movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[i]
        
    return movie_score

# Η συνάρτηση predict επιστρέφει δύο πίνακες:
# 1. real: πίνακας που περιέχει 0 ή 1 σε κάθε κελί ανάλογα με το αν θα έπρεπε να προταθεί η αντίστοιχη ταινία
# 2. pred: πίνακας που περιέχει 0 ή 1 σε κάθε κελί ανάλογα με το αν ο recommender προτείνει την αντίστοιχη ταινία

def predict(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: 
            pred.append(int(movie_score[movie] > 3)) # heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))  # έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
                                          # η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: 
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

In [15]:
train_ratings = pd.read_csv("train_ratings.csv")
test_ratings = pd.read_csv("test_ratings.csv")

In [16]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 0, 10)
count = 0
for movie in movie_score.items():
    if movie[1] > 3:
        count += 1
print("{} {}\n".format("Movies with score > 3: ", count))

print("§ Ενδεικτική αντιστοιχία §")
real, pred = predict(test_ratings, movie_score)

print("real:    ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(real[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(real[i], end=" ")
print()
print("predict: ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(pred[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(pred[i], end=" ")
print()
print()
print()

metrics = get_metrics(real, pred)
print("{}: {}%\n".format("Precision", round(metrics["precision"]*100,2)))
print("{}: {}%\n".format("Recall   ", round(metrics["recall"]*100,2)))
print("{}: {}%\n".format("F1       ", round(metrics["f1"]*100,2)))


Movies with score > 3:  166

§ Ενδεικτική αντιστοιχία §
real:    0 1 1 1 1 0 0 0 0 0 1 1 0 0 1 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 
predict: 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 


Precision: 62.5%

Recall   : 6.94%

F1       : 12.5%



In [17]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10, 20, movie_score)
count = 0
for movie in movie_score.items():
    if movie[1] > 3:
        count += 1
print("{} {}\n".format("Movies with score > 3: ", count))

print("§ Ενδεικτική αντιστοιχία §")
real, pred = predict(test_ratings, movie_score)

print("real:    ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(real[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(real[i], end=" ")
print()
print("predict: ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(pred[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(pred[i], end=" ")
print()
print()
print()

metrics = get_metrics(real, pred)
print("{}: {}%\n".format("Precision", round(metrics["precision"]*100,2)))
print("{}: {}%\n".format("Recall   ", round(metrics["recall"]*100,2)))
print("{}: {}%\n".format("F1       ", round(metrics["f1"]*100,2)))


Movies with score > 3:  1482

§ Ενδεικτική αντιστοιχία §
real:    0 1 1 1 1 0 0 0 0 0 1 1 0 0 1 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 
predict: 1 0 1 0 1 1 0 1 0 1 1 0 1 0 0 1 0 0 1 0 0 0 0 0 1 0 0 0 1 0 0 1 1 0 0 0 1 1 0 0 0 1 1 0 0 0 0 1 1 0 


Precision: 45.28%

Recall   : 33.33%

F1       : 38.4%



In [18]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 20, 30, movie_score)
count = 0
for movie in movie_score.items():
    if movie[1] > 3:
        count += 1
print("{} {}\n".format("Movies with score > 3: ", count))

print("§ Ενδεικτική αντιστοιχία §")
real, pred = predict(test_ratings, movie_score)

print("real:    ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(real[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(real[i], end=" ")
print()
print("predict: ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(pred[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(pred[i], end=" ")
print()
print()
print()

metrics = get_metrics(real, pred)
print("{}: {}%\n".format("Precision", round(metrics["precision"]*100,2)))
print("{}: {}%\n".format("Recall   ", round(metrics["recall"]*100,2)))
print("{}: {}%\n".format("F1       ", round(metrics["f1"]*100,2)))


Movies with score > 3:  1836

§ Ενδεικτική αντιστοιχία §
real:    0 1 1 1 1 0 0 0 0 0 1 1 0 0 1 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 
predict: 1 0 1 0 1 1 0 1 0 1 1 0 1 0 1 1 0 0 1 0 0 0 0 0 1 1 0 0 1 0 0 1 1 0 0 0 1 1 1 0 0 1 1 0 0 0 0 1 1 0 


Precision: 43.55%

Recall   : 37.5%

F1       : 40.3%



In [19]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30, 50, movie_score)
count = 0
for movie in movie_score.items():
    if movie[1] > 3:
        count += 1
print("{} {}\n".format("Movies with score > 3: ", count))

print("§ Ενδεικτική αντιστοιχία §")
real, pred = predict(test_ratings, movie_score)

print("real:    ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(real[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(real[i], end=" ")
print()
print("predict: ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(pred[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(pred[i], end=" ")
print()
print()
print()

metrics = get_metrics(real, pred)
print("{}: {}%\n".format("Precision", round(metrics["precision"]*100,2)))
print("{}: {}%\n".format("Recall   ", round(metrics["recall"]*100,2)))
print("{}: {}%\n".format("F1       ", round(metrics["f1"]*100,2)))


Movies with score > 3:  4038

§ Ενδεικτική αντιστοιχία §
real:    0 1 1 1 1 0 0 0 0 0 1 1 0 0 1 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 
predict: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 


Precision: 49.6%

Recall   : 86.11%

F1       : 62.94%



In [20]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights,start = 50, end = 70, movie_score=movie_score)
count = 0
for movie in movie_score.items():
    if movie[1] > 3:
        count += 1
print("{} {}\n".format("Movies with score > 3: ", count))

print("§ Ενδεικτική αντιστοιχία §")
real, pred = predict(test_ratings, movie_score)

print("real:    ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(real[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(real[i], end=" ")
print()
print("predict: ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(pred[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(pred[i], end=" ")
print()
print()
print()

metrics = get_metrics(real, pred)
print("{}: {}%\n".format("Precision", round(metrics["precision"]*100,2)))
print("{}: {}%\n".format("Recall   ", round(metrics["recall"]*100,2)))
print("{}: {}%\n".format("F1       ", round(metrics["f1"]*100,2)))


Movies with score > 3:  4505

§ Ενδεικτική αντιστοιχία §
real:    0 1 1 1 1 0 0 0 0 0 1 1 0 0 1 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 
predict: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 


Precision: 51.11%

Recall   : 95.83%

F1       : 66.67%



In [21]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights,start = 70, end = 90, movie_score=movie_score)
count = 0
for movie in movie_score.items():
    if movie[1] > 3:
        count += 1
print("{} {}\n".format("Movies with score > 3: ", count))

print("§ Ενδεικτική αντιστοιχία §")
real, pred = predict(test_ratings, movie_score)

print("real:    ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(real[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(real[i], end=" ")
print()
print("predict: ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(pred[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(pred[i], end=" ")
print()
print()
print()

metrics = get_metrics(real, pred)
print("{}: {}%\n".format("Precision", round(metrics["precision"]*100,2)))
print("{}: {}%\n".format("Recall   ", round(metrics["recall"]*100,2)))
print("{}: {}%\n".format("F1       ", round(metrics["f1"]*100,2)))


Movies with score > 3:  4547

§ Ενδεικτική αντιστοιχία §
real:    0 1 1 1 1 0 0 0 0 0 1 1 0 0 1 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 
predict: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 


Precision: 50.75%

Recall   : 94.44%

F1       : 66.02%



In [22]:
movie_score = train_recommender(train_ratings, rating_weights, score_weights,start = 90, movie_score=movie_score)
count = 0
for movie in movie_score.items():
    if movie[1] > 3:
        count += 1
print("{} {}\n".format("Movies with score > 3: ", count))

print("§ Ενδεικτική αντιστοιχία §")
real, pred = predict(test_ratings, movie_score)

print("real:    ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(real[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(real[i], end=" ")
print()
print("predict: ", end="")
for i in range(50):
    if real[i] == pred[i]:
        print("\x1b[32m", end="")
        print(pred[i], end=" ")
        print("\x1b[0m", end="")
    else:
        print(pred[i], end=" ")
print()
print()
print()

metrics = get_metrics(real, pred)
print("{}: {}%\n".format("Precision", round(metrics["precision"]*100,2)))
print("{}: {}%\n".format("Recall   ", round(metrics["recall"]*100,2)))
print("{}: {}%\n".format("F1       ", round(metrics["f1"]*100,2)))


Movies with score > 3:  4549

§ Ενδεικτική αντιστοιχία §
real:    0 1 1 1 1 0 0 0 0 0 1 1 0 0 1 1 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 
predict: 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 


Precision: 50.75%

Recall   : 94.44%

F1       : 66.02%



---
Σε καθένα απο τα τελευταία κελιά, ο recommender εκπαιδεύεται όλο και περισσότερο χρησιμοποιώντας τα ήδη υπάρχοντα δεδομένα του και εμπλουτίζοντάς τα περισσότερο με καινούργια (επιπλέον σειρές από το αρχείο με τα ratings του χρήστη). Παρατηρούμε πως μέχρι τα μισά κιόλας περίπου δεδομένα, η απόδοση του recommender τείνει να σταθεροποιηθεί, αν αυτή καθορίζεται απο τις μετρικές precision, recall και F1.

Γνωρίζουμε ότι:

$$precision = \frac{TP}{TP+FP}$$

$$recall    = \frac{TP}{TP+FN}$$

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

όπου TP: true positive, FP: false positive, FN: false negative

Παρατηρούμε, τόσο από τα ενδεικτικά αποτελέσματα που φαίνονται παραπάνω όσο και από ολόκληρους τους πίνακες με τα αποτελέσματα του recommender, ότι από ένα σημείο και μετά ο recommender προτείνει πολλές περισσότερες ταινίες από όσες θα έπρεπε. Μόλις χρησιμοποιηθούν όλα τα δεδομένα ratings τότε οι ταινίες που προτείνονται είναι 4549 (όσες έχουν συγκεντρώσει score πάνω από 3), δηλαδή σχεδόν όλες.

Με δεδομένο τον ορισμό των μετρικών, έτσι εξηγείται γιατί το precision συγκλίνει στο 50% (αν και σε αυτό συμβάλλει και η δομή των ίδιων των ratings), ενώ το recall στο 95%.

Τελικώς, ο recommender κρίνεται ότι είναι αρκετά χαμηλού επιπέδου, καθώς δεν είναι όσο επιλεκτικός θα θέλαμε. 

---